File: zigzag/zigzag.ts
 function convert(s: string, numRows: number): string {
    if(numRows === 1) return s
    const rows: string[] = new Array(numRows).fill('')
    let currentRow = 0
    let goingDown = true

    for (const letter of s) {
        rows[currentRow] += letter
        if(currentRow === 0) goingDown = true
        if(currentRow === numRows - 1) goingDown = false
        currentRow += goingDown ? 1 : -1
    }

    return rows.join('')
};

File: zigzag/zigzag.go
package main

import "strings"

func convert(s string, numRows int) string {
	if numRows == 1 {
		return s
	}
	rows := make([]string, numRows)
	goingDown := true
	currentRow := 0
	for _, char := range s {
		rows[currentRow] += string(char)
		if currentRow == 0 {
			goingDown = true
		}
		if currentRow == numRows-1 {
			goingDown = false
		}
		if goingDown {
			currentRow++
		} else {
			currentRow--
		}
	}
	return strings.Join(rows, "")
}
